File: goblin-core/src/main/java/org/goblinframework/core/serialization/fst/FstSerializer.java
package org.goblinframework.core.serialization.fst;

import org.goblinframework.api.annotation.Singleton;
import org.goblinframework.api.core.SerializerMode;
import org.goblinframework.core.exception.GoblinSerializationException;
import org.goblinframework.core.serialization.Serializer;
import org.jetbrains.annotations.NotNull;
import org.nustaq.serialization.FSTConfiguration;
import org.nustaq.serialization.FSTObjectInput;
import org.nustaq.serialization.FSTObjectOutput;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;

@Singleton
final public class FstSerializer implements Serializer {

  public static final FstSerializer INSTANCE = new FstSerializer();

  private FstSerializer() {
  }

  @NotNull
  @Override
  public SerializerMode mode() {
    return SerializerMode.FST;
  }

  @Override
  public void serialize(@NotNull Object obj, @NotNull OutputStream outStream) {
    if (!(obj instanceof Serializable)) {
      throw GoblinSerializationException.requiredSerializable(obj);
    }
    FSTConfiguration configuration = FstConfigurationFactory.INSTANCE.getConfiguration();
    try {
      FSTObjectOutput fo = configuration.getObjectOutput(outStream);
      fo.writeObject(obj);
      fo.flush();
    } catch (IOException ex) {
      throw new GoblinSerializationException(ex);
    }
  }

  @NotNull
  @Override
  public byte[] serialize(@NotNull Object obj) {
    if (!(obj instanceof Serializable)) {
      throw GoblinSerializationException.requiredSerializable(obj);
    }
    FSTConfiguration configuration = FstConfigurationFactory.INSTANCE.getConfiguration();
    try {
      FSTObjectOutput fo = configuration.getObjectOutput();
      fo.writeObject(obj);
      return fo.getCopyOfWrittenBuffer();
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    }
  }

  @NotNull
  @Override
  public Object deserialize(@NotNull InputStream inStream) {
    FSTConfiguration configuration = FstConfigurationFactory.INSTANCE.getConfiguration();
    try {
      FSTObjectInput fi = configuration.getObjectInput(inStream);
      return fi.readObject();
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    }
  }

  @NotNull
  @Override
  public Object deserialize(@NotNull byte[] bs) {
    FSTConfiguration configuration = FstConfigurationFactory.INSTANCE.getConfiguration();
    try {
      FSTObjectInput fi = configuration.getObjectInput(bs);
      return fi.readObject();
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    }
  }
}

File: goblinframework-core/src/main/java/org/goblinframework/core/serialization/fst/FstSerializer.kt
package org.goblinframework.core.serialization.fst

import org.goblinframework.core.serialization.Serializer
import org.goblinframework.core.serialization.SerializerMode
import java.io.InputStream
import java.io.OutputStream

class FstSerializer : Serializer {

  override fun mode(): SerializerMode {
    return SerializerMode.FST
  }

  override fun serialize(obj: Any, outStream: OutputStream) {
    val configuration = FstConfigurationFactory.configuration
    val fo = configuration.getObjectOutput(outStream)
    fo.writeObject(obj)
    fo.flush()
  }

  override fun serialize(obj: Any): ByteArray {
    val configuration = FstConfigurationFactory.configuration
    return configuration.asByteArray(obj)
  }

  override fun deserialize(inStream: InputStream): Any {
    val configuration = FstConfigurationFactory.configuration
    val fi = configuration.getObjectInput(inStream)
    return fi.readObject()
  }

  override fun deserialize(bs: ByteArray): Any {
    val configuration = FstConfigurationFactory.configuration
    return configuration.asObject(bs)
  }
}